[FEATURE] 이미지에서 인식된 가구와 선택된 가구의 교집합으로 조회하는 가구 카테고리 API

kbt82883 opened the pull request:
## 📣 Related Issue

- close #245 

## 📝 Summary

1. 생성된 이미지에서 객체 인식 모델로 인식된 가구들의 object365기반 영어단어를 요청으로 받아, 이미지 생성시 사용자가 선택했던 가구들간의 교집합 가구들을 가구 카테고리로 반환해야합니다.
2. 위의 과정으로 구해진 교집합의 가구 카테고리들은 생성된 이미지의 스타일에 따라 정렬되는 순서가 달라야합니다.

## 🙏 Question & PR point

로직이 조금 복잡해서, FurnitureServiceImplTest의 categories_intersection_sorted 테스트를 확인해보시면 이해하시는데 도움될 것 같습니다!

## 📬 Postman

<!-- postman 스크린샷을 첨부해주세요 -->

kbt82883 committed:
feat#247: FurnitureTag 매핑 클래스의 필드 추가

kbt82883 committed:
feat#245: Furniture클래스에 object365기반 단어 필드 추가

kbt82883 committed:
feat#245: houseId에 해당하는 가구 리스트 조회 쿼리 작성

kbt82883 committed:
feat#245: 생성된 이미지에서 인식된 가구 기반 가구 카테고리 조회 API 구현

kbt82883 committed:
feat#245: 응답 DTO 작성

kbt82883 committed:
test#245: 비즈니스로직 테스트코드 작성

kbt82883 committed:
fix#245: API 엔드포인트 수정

kbt82883 committed:
Merge branch 'develop' into feat/#245/furniture-category-from-image

PBEM22 commented:
### 📝 Code Coverage
|Overall Project|66.06% **`-0.5%`**|:x:|
|:-|:-|:-:|
|Files changed|70.69%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[FurnitureCategoriesResponse.java](https://github.com/TEAM-HOUME/HOUME-SERVER/blob/fa5ecf2edf89db1c6186a40b92bb11429f1c9def/src%2Fmain%2Fjava%2For%2Fsopt%2Fhoume%2Fdomain%2Ffurniture%2Fdto%2Fresponse%2FFurnitureCategoriesResponse.java)|100%|:green_apple:|
|[FurnitureServiceImpl.java](https://github.com/TEAM-HOUME/HOUME-SERVER/blob/fa5ecf2edf89db1c6186a40b92bb11429f1c9def/src%2Fmain%2Fjava%2For%2Fsopt%2Fhoume%2Fdomain%2Ffurniture%2Fservice%2FFurnitureServiceImpl.java)|89.42%|:green_apple:|
|[Furniture.java](https://github.com/TEAM-HOUME/HOUME-SERVER/blob/fa5ecf2edf89db1c6186a40b92bb11429f1c9def/src%2Fmain%2Fjava%2For%2Fsopt%2Fhoume%2Fdomain%2Ffurniture%2Fentity%2FFurniture.java)|88.57%|:green_apple:|
|[FurnitureTag.java](https://github.com/TEAM-HOUME/HOUME-SERVER/blob/fa5ecf2edf89db1c6186a40b92bb11429f1c9def/src%2Fmain%2Fjava%2For%2Fsopt%2Fhoume%2Fdomain%2Ffurniture%2Fentity%2FFurnitureTag.java)|73.57% **`-4.29%`**|:x:|
|[FurnitureController.java](https://github.com/TEAM-HOUME/HOUME-SERVER/blob/fa5ecf2edf89db1c6186a40b92bb11429f1c9def/src%2Fmain%2Fjava%2For%2Fsopt%2Fhoume%2Fdomain%2Ffurniture%2Fcontroller%2FFurnitureController.java)|23.08% **`-46.15%`**|:x:|
|[FurnitureCustomRepositoryImpl.java](https://github.com/TEAM-HOUME/HOUME-SERVER/blob/fa5ecf2edf89db1c6186a40b92bb11429f1c9def/src%2Fmain%2Fjava%2For%2Fsopt%2Fhoume%2Fdomain%2Ffurniture%2Frepository%2FFurnitureCustomRepositoryImpl.java)|8.96% **`-49.25%`**|:x:|

PBEM22 committed:
refactor: #256 이미지 생성 Request 구조 변경 (bedId 삭제) (#257)

# Conflicts:
#	src/main/java/or/sopt/houme/domain/furniture/service/FurnitureService.java
#	src/main/java/or/sopt/houme/domain/furniture/service/FurnitureServiceImpl.java

gdbs1107 commented on src/main/java/or/sopt/houme/domain/furniture/repository/FurnitureCustomRepositoryImpl.java:
```diff
@@ -36,4 +38,19 @@ public List<Furniture> findAllWithTags() {
-}
\ No newline at end of file
+
+    @Override
+    public List<Furniture> findAllByHouseId(Long houseId) {
+        QHouse house = QHouse.house;
+        QHouseFurniture houseFurniture = QHouseFurniture.houseFurniture;
+        QFurniture furniture = QFurniture.furniture;
+
+        return queryFactory
+                .select(furniture)
+                .from(houseFurniture)
+                .join(houseFurniture.furniture, furniture)
+                .join(houseFurniture.house, house)
+                .where(house.id.eq(houseId))
+                .fetch();
+    }
```
쿼리를 직접 확인해보진 않았지만 N+1이 발생 할 수 있을 것 같아요! 추후 리팩터링이 필요 할 것 같습니다

kbt82883 replied:
네 추후 쿼리 리팩터링하면서 함께 처리하겠습니다!

gdbs1107 reviewed (approved):
깔끔하고 좋은 리팩터링인 것 같습니다!

충돌만 해결하고 병합해주시면 좋을 것 같아요

kbt82883 committed:
feat#245: 교집합 빈 리스트인 경우 반환 처리

kbt82883 committed:
docs#245: 추후 리팩터링 필요 주석 추가

kbt82883 committed:
refactor#245: 사용하지 않는 import문 제거

kbt82883 committed:
fix#245: 충돌 해결

# Conflicts:
#	src/main/java/or/sopt/houme/domain/furniture/service/FurnitureService.java
#	src/main/java/or/sopt/houme/domain/furniture/service/FurnitureServiceImpl.java

kbt82883 committed:
test#245: 테스트 코드 작성

kbt82883 committed:
delete#245: 테스트 코드 삭제

kbt82883 committed:
delete#245: 중복 코드 삭제

kbt82883 committed:
Merge branch 'develop' into feat/#245/furniture-category-from-image

kbt82883 committed:
fix#245: 매핑 객체에서 중복된 가구 조회 처리
